File: self_prop.py
import random
import numpy as np
import matplotlib.pyplot as plt
import math
from itertools import combinations
import copy
import csv
from sys import getsizeof
import pandas as pd
import time

class Particle:
    def __init__(self, pos, linvel):
        self.pos=pos
        self.linvel=linvel
        # self.isFromWedge=isFromWedge

class Chain:
    def __init__(self, chainSize=None, pos=None, angle=None, isWedge=False):    
        if pos is None:
            self.pos=np.array([random.uniform(boxX[0], boxX[1]), random.uniform(boxY[0], boxY[1])], dtype=np.float64)
        else:
            self.pos=pos

        if angle is None:
            self.angle=random.uniform(0,2*np.pi)
        else:
            self.angle=angle
        self.orient= np.array([np.cos(self.angle), np.sin(self.angle)], dtype=np.float64)      #unit vector, equal to distance between COM of consecutive particles
        self.linvel=np.array([0,0], dtype=np.float64)
        self.angvel=0                                                                          #vector-wise this is in z
        if chainSize is None:
            chainSize=globalChainSize
        self.chainSize=chainSize
        self.force=Fo*self.orient
        self.moment=0
        self.particles=[]
        self.isWedge=isWedge
        if self.chainSize%2!=0:
            for i in range(-1*int(self.chainSize/2),int(self.chainSize/2)+1):
                particlePos=self.pos+(d*i*self.orient)
                particleVel=self.linvel+zxycross(self.angvel, particlePos-self.pos)
                self.particles.append(Particle(particlePos, particleVel))
        else:
            for i in range(-1*(int(self.chainSize/2)-1),int(self.chainSize/2)+1):
                particlePos=self.pos+(d*i*self.orient)-(d/2)*self.orient
                particleVel=self.linvel+zxycross(self.angvel, particlePos-self.pos)
                self.particles.append(Particle(particlePos, particleVel))
                
    def updateParticles(self):
        for i in range(len(self.particles)):
            self.particles[i].pos=self.pos+((i-int(len(self.particles)/2))*d*self.orient)
            self.particles[i].linvel=self.linvel+zxycross(self.angvel, self.particles[i].pos-self.pos)

# cross of a vector in x-y plane with a vector perpendicular to it
def zxycross(z:float, xy:np.ndarray):
    return np.array([-1*xy[1]*z, xy[0]*z])

# force on a due to b
def getForce(a:Particle, b:Particle)->np.ndarray:
    # xdist = a.pos[0]-b.pos[0]
    # ydist = a.pos[1]-b.pos[1]
    # if abs(xdist)>boxSize[0]-abs(xdist):
    #     xdist*=-1
    # if abs(ydist)>boxSize[1]-abs(ydist):
    #     ydist*=-1
    # xdist = min(xdist, boxSize[0]-xdist)
    # ydist = min(ydist, boxSize[1]-ydist)
    # rmod=np.linalg.norm((xdist, ydist))
    apos = copy.deepcopy(a.pos)
    bpos= copy.deepcopy(b.pos)
    if abs(apos[0]-bpos[0])>(boxSize[0])/2:
        if apos[0]>bpos[0]:
            bpos[0]+=boxSize[0]
        else:
            apos[0]+=boxSize[0]

    if abs(apos[1]-bpos[1])>(boxSize[1])/2:
        if apos[1]>bpos[1]:
            bpos[1]+=boxSize[1]
        else:
            apos[1]+=boxSize[1]

    rmod=np.linalg.norm(apos-bpos)
    rhat=(apos-bpos)/rmod
    f = ((Uo*math.exp(-1*rmod/lamb))/rmod)*((1/rmod)+(1/lamb))*rhat
    return f
    

def updateChainForces(a:Chain, b:Chain):
    # if not isinstance(a, Chain) or not isinstance(b, Chain):
    #     raise TypeError("force cannot be calculated for non chains")
    if (not a.isWedge) and (not b.isWedge):
        xdist = abs(a.pos[0]-b.pos[0])
        ydist = abs(a.pos[1]-b.pos[1])
        # if boxSize[0]-xdist <=0:
        #     raise ValueError("box size too small")
        if min(xdist, boxSize[0]-xdist)>2*l or min(ydist, boxSize[1]-ydist)>2*l:
            # print("skipped")
            return

    for i in range(len(a.particles)):
        for j in range(len(b.particles)):
            force=getForce(a.particles[i], b.particles[j])
            a.force+=force
            b.force-=force
            a.moment+=np.cross(a.particles[i].pos-a.pos, force)
            b.moment+=np.cross(b.particles[j].pos-b.pos, -1*force)

def updateChainPositions(chain:Chain):
    # if not isinstance(chain, Chain):
    #     raise TypeError("force cannot be calculated for non chains")
    chain.pos+=chain.linvel*timeStep
    chain.pos[0]=((chain.pos[0]-boxX[0])%(boxSize[0]))+boxX[0]
    chain.pos[1]=((chain.pos[1]-boxY[0])%(boxSize[1]))+boxY[0]
    chain.angle+=chain.angvel*timeStep
    if chain.angle>2*np.pi:
        chain.angle-=2*np.pi
    if chain.angle<0:
        chain.angle+=2*np.pi
    chain.orient=np.array([np.cos(chain.angle), np.sin(chain.angle)])

def updateChainVelocities(chain:Chain):
    # if not isinstance(chain, Chain):
    #     raise TypeError("force cannot be calculated for non chains")
    chain.linvel=((np.dot(chain.force, chain.orient)/ft1)*chain.orient)+((chain.force-np.dot(chain.force, chain.orient)*chain.orient)/ft2)
    chain.angvel=chain.moment/fr

def getFrictionCoeff():
    a=aRatio
    fParallel = 2*np.pi/(np.log(a)-0.207+(0.98/a)-(0.133/(a**2)))
    fPerpendicular = 2*np.pi/(np.log(a)+0.839+(0.185/a)+(0.233/(a**2)))
    fRot = np.pi*(a**2)/(3*(np.log(a)-0.662+(0.917/a)-(0.05/(a**2))))
    return fo*fParallel, fo*fPerpendicular, fo*fRot

def addWedge(chainArray):
    # if wedgeSize%2==0:
    #     raise ValueError("length of wedge must be odd")
    upperChain = Chain(chainSize=wedgeSize, pos=np.array(([(wedgeSize-1)*d*np.cos(wedgeAngle/2)/2,(wedgeSize-1)*d*np.sin(wedgeAngle/2)/2]), dtype=np.float64), 
                       angle=wedgeAngle/2, isWedge=True)
    
    lowerChain = Chain(chainSize=wedgeSize, pos=np.array(([(wedgeSize+1)*d*np.cos(-1*wedgeAngle/2)/2,(wedgeSize+1)*d*np.sin(-1*wedgeAngle/2)/2]), dtype=np.float64), 
                       angle=-1*wedgeAngle/2, isWedge=True)
    
    chainArray.append(upperChain)
    chainArray.append(lowerChain)
    # return upperChain, lowerChain



if __name__=="__main__":
    Uo=1
    lamb=1
    l=10
    aRatio=l/lamb
    globalChainSize=int(round(9*aRatio/8))
    d=l/(math.sqrt((globalChainSize+1)*(globalChainSize-1)))
    timeStep=0.05
    Fo=2
    chainNos=30
    fo=1
    iterations=50
    wedgeSize=globalChainSize*4
    wedgeAngle=np.pi/2
    ft1, ft2, fr = getFrictionCoeff()
    boxX=(-5*l,5*l)
    boxY=(-5*l,5*l)
    boxSize = (boxX[1]-boxX[0], boxY[1]-boxY[0])
    totalChainNos=chainNos+2 #including wedges
    
    chains=[]
    addWedge(chains)
    for i in range(chainNos):
        chains.append(Chain())

    dataArr=np.zeros([iterations, totalChainNos, 6])
    for t in range(0, iterations):
        # timeData=[[t]]
        print(t)
        t_init=time.time()
        comb = combinations(chains, 2)

        for i in list(comb):
            updateChainForces(i[0], i[1])
        for i in range(len(chains)):
            # print(i)
            chain=chains[i]
            if not chain.isWedge:
                updateChainVelocities(chain)
                updateChainPositions(chain)
                chain.updateParticles()
            chain.force=Fo*chain.orient
            chain.moment=0

            dataArr[t][i][0]=chain.pos[0]
            dataArr[t][i][1]=chain.pos[1]
            dataArr[t][i][2]=chain.angle
            dataArr[t][i][3]=chain.linvel[0]
            dataArr[t][i][4]=chain.linvel[1]
            dataArr[t][i][5]=chain.angvel
            
            for j in range(len(chain.particles)):
                particle=chain.particles[j]
                if chain.isWedge:
                    plt.plot(particle.pos[0], particle.pos[1], 'b', marker=".", markersize=6)
                else:
                # for j in range(len(chain.particles)):
                    plt.plot(particle.pos[0], particle.pos[1], 'r', marker=".", markersize=6)

        plt.xlim(boxX[0], boxX[1])
        plt.ylim(boxY[0], boxY[1])

        plt.savefig("./out/test"+str(t)+".png")
        plt.clf()
        print(time.time()-t_init)
    dataArr=dataArr.reshape([iterations, totalChainNos*6])
    df=pd.DataFrame(dataArr)
    df.to_csv("./out.csv",header=False)

File: self_prop.cpp
#include <cmath>
#include <iostream>
#include <fstream>
#include <filesystem>
using namespace std;

const float FLPI = M_PI;
const float Uo = 1;
const float lamb = 1;
const float l = 10;
const float aRatio = l/lamb;
const int globalChainSize = int(round(9*aRatio/8));
const float d = l/(sqrt((globalChainSize+1)*(globalChainSize-1)));
const float timeStep = 0.001;
const float Fo = 1;
const int chainNos = 100;
const float fo = 1;
const int wedgeSize=globalChainSize*10;
const float wedgeAngle=9.0*(FLPI/18);
const float chain_chain_cutoff=l+2*(lamb);
const float chain_wedge_cutoff=l/2+5*l+(lamb);
const int saveChunkSize=100;
const int iter=1000000;
const int saveFreq=1000;
const int mulFac=11;
const int kappa=0.001;
// const float rootpiinv=0.56418958354;
// float newpos[2];
// float newangle;
const float quartmax=1.0*RAND_MAX/4;
const string csvName="./friction/test0001.csv";
std::ofstream csvFile;


float* zxycross(float z, float* xy){
        float* tmp = new float[2];
        tmp[0] = -1*xy[1]*z, tmp[1] = xy[0]*z;
        return tmp;
}

float dot(float* x, float* y){
        float tmp = x[0]*y[0]+x[1]*y[1];
        return tmp;
}

float newmod(float a, float b){
        float mod_val = fmod(a, b);
        return mod_val >= 0 ? mod_val : b + mod_val;
}


float* getFrictionCoeff(){
        float* friction = new float[3];
        friction[0] =  2*FLPI/(log(aRatio)-0.207+(0.98/aRatio)-(0.133/(aRatio*aRatio)));
        friction[1] =  2*FLPI/(log(aRatio)+0.839+(0.185/aRatio)+(0.233/(aRatio*aRatio)));
        friction[2] =  FLPI*(aRatio*aRatio)/(3*(log(aRatio)-0.662+(0.917/aRatio)-(0.05/(aRatio*aRatio))));
        return friction;
}

const float* friction = getFrictionCoeff();
const float boxX[2] = {-15*l, 15*l};
const float boxY[2] = {-15*l, 15*l};

class Particle{
        public:
                float pos[2];
                float linvel[2];
};
class SPR{
        public:
                float pos[2];
                // float prevpos[2];
                float linvel[2];
                float orient[2];
                float force[2];
                float angle;
                // float prevangle;
                float angvel;
                float moment;
                Particle particles[globalChainSize];
                void update_particle(){
                        if (globalChainSize%2!=0){
                                for (int i=-1*int(globalChainSize/2); i<=int(globalChainSize/2); i++){
                                        int j=i;
                                        if (j<0){j+=globalChainSize;}
                                        particles[j].pos[0] = pos[0]+d*i*orient[0];
                                        particles[j].pos[1] = pos[1]+d*i*orient[1];
                                        float* rCOM= new float[2];
                                        rCOM[0] = particles[j].pos[0]-pos[0], rCOM[1] =  particles[j].pos[1]-pos[1];
                                        float* cross=zxycross(angvel, rCOM);
                                        particles[j].linvel[0] = linvel[0]+cross[0];
                                        particles[j].linvel[1] = linvel[1]+cross[1];
                                        delete[] cross; delete[] rCOM;
                                }
                        }
                        else{
                                for (int i=-1*(int(globalChainSize/2)-1); i<=int(globalChainSize/2); i++){
                                        int j=i;
                                        if (j<0){j+=globalChainSize;}
                                        particles[j].pos[0] = pos[0]+d*i*orient[0]-d*orient[0]/2;
                                        particles[j].pos[1] = pos[1]+d*i*orient[1]-d*orient[1]/2;
                                        float* rCOM= new float[2];
                                        rCOM[0] = particles[j].pos[0]-pos[0], rCOM[1] =  particles[j].pos[1]-pos[1];
                                        float* cross=zxycross(angvel, rCOM);
                                        particles[j].linvel[0] = linvel[0]+cross[0];
                                        particles[j].linvel[1] = linvel[1]+cross[1];
                                        delete[] cross; delete[] rCOM;
                                }
                        }
                }

                //Method to update chain parameters given forces are updated already. Call at every timestep
                void update_chain(){
                        linvel[0] = dot(force, orient)*orient[0]/friction[0]+(force[0]-dot(force, orient)*orient[0])/friction[1];
                        linvel[1] = dot(force, orient)*orient[1]/friction[0]+(force[1]-dot(force, orient)*orient[1])/friction[1];
                        // if(abs(linvel[0])<0.01) linvel[0]=0;
                        // if(abs(linvel[1])<0.01) linvel[1]=0;
                        angvel = moment/friction[2];
                        // if(abs(angvel)<0.0001) angvel=0;
                        // newpos[0] = (2*timeStep*linvel[0])+prevpos[0];
                        // newpos[1] = (2*timeStep*linvel[1])+prevpos[1];
                        // prevpos[0]=pos[0]; prevpos[1]=pos[1];
                        // pos[0]=newpos[0]; pos[1]=newpos[1];
                        pos[0]+=linvel[0]*timeStep;
                        pos[1]+=linvel[1]*timeStep;
                        pos[0]=newmod((pos[0]-boxX[0]), (boxX[1]-boxX[0]))+boxX[0];
                        pos[1]=newmod((pos[1]-boxY[0]), (boxY[1]-boxY[0]))+boxY[0];
                        // newangle = (2*timeStep*angvel)+prevangle;
                        // prevangle = angle; angle = newangle;
                        angle+=angvel*timeStep;
                        if (angle>2*FLPI){
                                angle-=2*FLPI;
                        }
                        if (angle<0){
                                angle+=2*FLPI;
                        }
                        orient[0] = cos(angle), orient[1] = sin(angle);
                }

};
class Wedge{
        public:
                float pos[2];
                float orient[2];
                Particle particles[wedgeSize];
                void update_Wedge(){
                        if (wedgeSize%2!=0){
                                for (int i=-1*int(wedgeSize/2); i<=int(wedgeSize/2); i++){
                                        int j=i;
                                        if (j<0){j+=wedgeSize;}
                                        particles[j].pos[0] = pos[0]+d*i*orient[0];
                                        particles[j].pos[1] = pos[1]+d*i*orient[1];
                                        particles[j].linvel[0] = 0;
                                        particles[j].linvel[1] = 0;
                                }
                        }
                        else{
                                for (int i=-1*(int(wedgeSize/2)-1); i<=int(wedgeSize/2); i++){
                                        int j=i;
                                        if (j<0){j+=wedgeSize;}
                                        particles[j].pos[0] = pos[0]+d*i*orient[0]-d*orient[0]/2;
                                        particles[j].pos[1] = pos[1]+d*i*orient[1]-d*orient[1]/2;
                                        particles[j].linvel[0] = 0;
                                        particles[j].linvel[1] = 0;
                                }
                        }
                }
};

float* getForce (Particle a, Particle b){
        float* tmp = new float[2];
        tmp[0]=0, tmp[1]=0; 
        if (abs(a.pos[0]-b.pos[0])>(boxX[1]-boxX[0])/2){
                (a.pos[0]>b.pos[0])?(b.pos[0]+=boxX[1]-boxX[0]):(a.pos[0]+=boxX[1]-boxX[0]);
        }
        if (abs(a.pos[1]-b.pos[1])>(boxY[1]-boxY[0])/2){
                (a.pos[1]>b.pos[1])?(b.pos[1]+=boxY[1]-boxY[0]):(a.pos[1]+=boxY[1]-boxY[0]);
        }
        float rRel[2] = {a.pos[0]-b.pos[0], a.pos[1]-b.pos[1]};
        float rmod = sqrt(dot(rRel, rRel));

        // Particle-particle friction
        if(rmod<=1.5*lamb){
                float rhat[2]={rRel[0]/rmod, rRel[1]/rmod};
                tmp[0]=a.linvel[0]-b.linvel[0];
                tmp[1]=a.linvel[1]-b.linvel[1];
                float dotprod=dot(tmp, rhat);
                tmp[0]=-1*kappa*(tmp[0]-(dotprod*rhat[0]));
                tmp[1]=-1*kappa*(tmp[1]-(dotprod*rhat[1]));
        }
        // Lennard Jones
        // float fmod = (Uo/pow(rmod,3));
        // tmp[0] = fmod*rRel[0];
        // tmp[1] = fmod*rRel[1];

        // Yukawa
        float fmod = ((Uo*exp(-1*rmod/lamb))/rmod)*((1/rmod)+(1/lamb))/rmod;
        tmp[0] = tmp[0]+((Uo*exp(-1*rmod/lamb))/rmod)*((1/rmod)+(1/lamb))*rRel[0]/rmod;
        tmp[1] = tmp[0]+((Uo*exp(-1*rmod/lamb))/rmod)*((1/rmod)+(1/lamb))*rRel[1]/rmod;
        return tmp;
}
void SPR_SPR(SPR& a, SPR& b){
	if (abs(a.pos[0]-b.pos[0])>chain_chain_cutoff or abs(a.pos[1]-b.pos[1])>chain_chain_cutoff){return;}
        for (int i=0; i<globalChainSize; i++){
                for (int j=0; j<globalChainSize; j++){
                        float* frc;
                        frc = getForce(a.particles[i], b.particles[j]);
                        a.force[0]+=frc[0]; a.force[1]+=frc[1];
                        b.force[0]-=frc[0]; b.force[1]-=frc[1];
                        a.moment+=(a.particles[i].pos[0]-a.pos[0])*frc[1]-(a.particles[i].pos[1]-a.pos[1])*frc[0];
                        b.moment+=-(b.particles[i].pos[0]-b.pos[0])*frc[1]+(b.particles[i].pos[1]-b.pos[1])*frc[0];
                        delete[] frc;
                }
        }
}
void SPR_Wedge(SPR& a, Wedge& b){
	if (abs(a.pos[0]-b.pos[0])>chain_wedge_cutoff or abs(a.pos[1]-b.pos[1])>chain_wedge_cutoff) return;
        for (int i=0; i<globalChainSize; i++){
                for (int j=0; j<wedgeSize; j++){
                        float* frc;
                        frc = getForce(a.particles[i], b.particles[j]);
                        frc[0]*=mulFac; frc[1]*=mulFac;
                        a.force[0]+=frc[0]; a.force[1]+=frc[1];
                        a.moment+=(a.particles[i].pos[0]-a.pos[0])*frc[1]-(a.particles[i].pos[1]-a.pos[1])*frc[0];
                        delete[] frc;
                }
        }
}

void addData(float arr[], SPR& a, int t, int i){
    int tOff = t*3*chainNos;
    int chainOff=i*3;
    int totalOff=tOff+chainOff;
    arr[totalOff]=a.pos[0];
    arr[totalOff+1]=a.pos[1];
    arr[totalOff+2]=a.angle;
//     arr[totalOff+3]=a.linvel[0];
//     arr[totalOff+4]=a.linvel[1];
//     arr[totalOff+5]=a.angvel;
}

void saveData(float arr[]){
    int tOff=0, chainOff=0, totalOff=0;
    csvFile.open(csvName, ios::app);
    for (int t=0; t<saveChunkSize; t++){
        tOff = t*3*chainNos;
        for (int i=0; i<chainNos; i++){
            chainOff=i*3;
            totalOff=tOff+chainOff;
            csvFile << arr[totalOff] << "," << arr[totalOff+1] << "," << arr[totalOff+2] << ",";              
        }
        csvFile << endl;
    }
    csvFile.close();

}

void initCSV(){
        remove(csvName.c_str());
        csvFile.open(csvName, ios::app);
        csvFile<<"timeStep,iter,Uo,Fo,chainNos,wedgeSize,wedgeAngle,saveFreq"<<endl;
        csvFile<<timeStep<<","<<iter<<","<<Uo<<","<<Fo<<","<<chainNos<<","<<wedgeSize<<","<<wedgeAngle<<","<<saveFreq<<endl;
        csvFile.close(); 
}

float getRandMoment(){
        float uni = (1.0*rand()/quartmax)-2;
        float ret= ((uni>=0)-(uni<0))*30*exp(-1*pow(uni,2));
        return ret;
}


int main(){
        time_t start,end;
        time(&start);
        SPR Rods[chainNos];
        float csvData[chainNos*3*saveChunkSize];
        for (int i=0; i<chainNos; i++){
                Rods[i].pos[0]=boxX[0]+(i%10)*11.13, Rods[i].pos[1]=boxY[0]+(i/10)*11.13;
                Rods[i].linvel[0]=0, Rods[i].linvel[1]=0;
                Rods[i].angle=i;
                // Rods[i].prevangle=i;
                Rods[i].angvel=0;
                // Rods[i].prevpos[0]=boxX[0]+(i%30)*11.13, Rods[i].prevpos[1]=boxY[0]+(i/30)*11.13;
                Rods[i].orient[0]=cos(Rods[i].angle); Rods[i].orient[1]=sin(Rods[i].angle);
                Rods[i].force[0]=Fo*Rods[i].orient[0]; Rods[i].force[1]=Fo*Rods[i].orient[1];
                Rods[i].moment=0;
                Rods[i].update_particle();
        }
	Wedge lowerWedge, upperWedge;
        lowerWedge.orient[0]=cos(-1*wedgeAngle/2); lowerWedge.orient[1]=sin(-1*wedgeAngle/2);
        upperWedge.orient[0]=cos(wedgeAngle/2); upperWedge.orient[1]=sin(wedgeAngle/2);
        lowerWedge.pos[0]=(wedgeSize+1)*d*lowerWedge.orient[0]/2; lowerWedge.pos[1]=(wedgeSize+1)*d*lowerWedge.orient[1]/2;
        upperWedge.pos[0]=(wedgeSize-1)*d*upperWedge.orient[0]/2; upperWedge.pos[1]=(wedgeSize-1)*d*upperWedge.orient[1]/2;
        lowerWedge.update_Wedge(); upperWedge.update_Wedge();
        
        initCSV();
        for (int t=0; t<iter; t++){
                for (int i=0; i<chainNos; i++){
                        for (int j=i+1; j<chainNos; j++){
                                SPR_SPR(Rods[i], Rods[j]);
                        }
			SPR_Wedge(Rods[i], lowerWedge);
                        SPR_Wedge(Rods[i], upperWedge);
                        Rods[i].update_chain();
                        Rods[i].update_particle();
                        Rods[i].force[0]=Fo*Rods[i].orient[0]; Rods[i].force[1]=Fo*Rods[i].orient[1];
                        Rods[i].moment=0;
                        if (t%saveFreq==0) addData(csvData, Rods[i], (t/saveFreq)%saveChunkSize, i);
                }
                if (t%(saveChunkSize*saveFreq)==(saveFreq*saveChunkSize)-1){
                    
                    saveData(csvData);
                }
                // cout<<t<<endl;
                // cout<<t<<endl;
                // cout << Rods[0].pos[0]; cout << ","; cout << Rods[0].linvel[0]; cout <<","; cout << Rods[0].angle;
                // cout<<"\n";
        }
        time(&end);
        cout<<"TIME: "<<double(end-start);
        return 0;
}
